### PR TITLE
Allow users to change version slug

### PR DESCRIPTION
**Pull request by stsewd**
Machine created versions aren't allowed to be changed,
since we use their harcoded slug to identify them as latest and stable.

Ref #5318

### CODE REVIEW

File: readthedocs/builds/forms.py [hunk 1]
```diff
@@ -14,7 +17,37 @@ class VersionForm(HideProtectedLevelMixin, forms.ModelForm):
 
     class Meta:
         model = Version
-        fields = ['active', 'privacy_level']
+        fields = ['slug', 'active', 'privacy_level']
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.fields['slug'].help_text = 'Warning: changing the slug will break existing URLs'
```
Comment (humitos) [inline readthedocs/builds/forms.py]: This may need to be translated.

File: docs/faq.rst [hunk 2]
```diff
@@ -24,14 +24,17 @@ as it `breaks the internet <http://www.w3.org/Provider/Style/URI.html>`_.
...
 How do I change the version slug of my project?
 -----------------------------------------------
 
-We don't support allowing folks to change the slug for their versions.
-But you can rename the branch/tag to achieve this.
-If that isn't enough,
-you can request the change sending an email to support@readthedocs.org.
+One way is renaming the branch/tag of your version.
+But if that isn't possible,
+you can rename the slug from the :guilabel:`Versions` tab.
```
Comment (humitos) [inline docs/faq.rst]:
```suggestion
You can achieve this by renaming the branch/tag of your version.
If that is not possible,
you can change the slug of the version from the :guilabel:`Versions` tab.
```

File: docs/faq.rst [hunk 3]
```diff
@@ -24,14 +24,17 @@ as it `breaks the internet <http://www.w3.org/Provider/Style/URI.html>`_.
...
+
+.. warning::
+
+   Be careful, changing the slug will break existing URLs.
```
Comment (humitos) [inline docs/faq.rst]: Can we suggest a good redirect to avoid this? Like, "You may want to add an Exact Redirect from `/en/oldversion/$rest` to `/en/newversion/$rest`" or similar.

Comment (humitos) [reply docs/faq.rst]: I think the warning message should still be there, though.

File: readthedocs/builds/forms.py [hunk 4]
```diff
@@ -14,7 +17,37 @@ class VersionForm(HideProtectedLevelMixin, forms.ModelForm):
...
+    def clean_slug(self):
+        slugifier = VersionSlugify(
+            ok_chars=VERSION_OK_CHARS,
+            test_pattern=VERSION_TEST_PATTERN,
+        )
+        original_slug = self.cleaned_data.get('slug')
+
+        slug = slugifier.slugify(original_slug, check_pattern=False)
+        if not slugifier.is_valid(slug):
+            msg = _('The slug "{slug}" is not valid.')
```
Comment (humitos) [inline readthedocs/builds/forms.py]: Can we communicate which are the valid chars here?

File: readthedocs/builds/version_slug.py [hunk 5]
```diff
@@ -50,36 +48,23 @@ def get_fields_with_model(cls):
 #   (?: ... ) -- wrap everything so that the pattern cannot escape when used in
 #                regexes.
 VERSION_SLUG_REGEX = '(?:[a-z0-9A-Z][-._a-z0-9A-Z]*?)'
+VERSION_OK_CHARS = '-._'  # dash, dot, underscore
+VERSION_TEST_PATTERN = re.compile('^{pattern}$'.format(pattern=VERSION_SLUG_REGEX))
+VERSION_FALLBACK_SLUG = 'unknown'
 
 
-class VersionSlugField(models.CharField):
+class VersionSlugify:
```
Comment (humitos) [inline readthedocs/builds/version_slug.py]: I understand that this new class is just a refactor to get out some hardcoded code from the VersionSlugField to be able to use it from the Form validation as well. Am I right?

In case that you change some logic, can you pointed to me? I didn't really find any.

File: readthedocs/builds/version_slug.py [hunk 6]
```diff
@@ -94,18 +79,23 @@ def _normalize(self, content):
...
+        return self.test_pattern.match(content)
+
+    def slugify(self, content, check_pattern=True):
         """
         Make ``content`` a valid slug.
 
         It uses ``unicode-slugify`` behind the scenes which works properly with
         Unicode characters.
+
+        If `check_pattern` is `True`, it checks that the final slug is valid.
```
Comment (humitos) [inline readthedocs/builds/version_slug.py]: It may mention here that it raises an exception if it's not valid.

File: readthedocs/builds/version_slug.py [hunk 7]
```diff
@@ -114,13 +104,40 @@ def slugify(self, content):
...
 
-        # Remove first character wile it's an invalid character for the
+        # Remove first character while it's an invalid character for the
         # beginning of the slug
-        slugified = slugified.lstrip(self.ok_chars)
+        slug = slug.lstrip(self.ok_chars)
+        slug = slug or self.fallback_slug
+
+        if check_pattern and not self.is_valid(slug):
+            raise Exception(f'Invalid generated slug: {slug}')
```
Comment (humitos) [inline readthedocs/builds/version_slug.py]: We may want to create our own exception for this: `InvalidVersionSlug` or similar.

File: readthedocs/builds/version_slug.py [hunk 8]
```diff
@@ -94,18 +79,23 @@ def _normalize(self, content):
         """
         return re.sub('[/%!?]', '-', content)
 
-    def slugify(self, content):
+    def is_valid(self, content):
+        return self.test_pattern.match(content)
+
+    def slugify(self, content, check_pattern=True):
```
Comment (humitos) [inline readthedocs/builds/version_slug.py]: Why did you add the `check_pattern` if we are using it always in `False`? It didn't exist before and now it is changing the default behavior to raise an exception.

File: readthedocs/rtd_tests/tests/test_build_forms.py [hunk 9]
```diff
@@ -43,10 +42,135 @@ def test_default_version_is_inactive(self):
...
+        # Form works
+        form = VersionForm(
+            {
+                'slug': 'latest',
+                'active': False,
+                'privacy_level': PUBLIC,
+            },
+            instance=version,
+        )
+        self.assertTrue(form.is_valid())
```
Comment (humitos) [inline readthedocs/rtd_tests/tests/test_build_forms.py]: Why the form is valid if the version has `machine=True`?

File: readthedocs/builds/version_slug.py (hunk 5 above)
Comment (stsewd) [reply readthedocs/builds/version_slug.py]: Yeah, it just extracts the code to make it reusable

File: readthedocs/rtd_tests/tests/test_build_forms.py (hunk 9 above)
Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_build_forms.py]: Django doesn't raise an exception if the value is changed (when disable=False), django just ignores the value, since users aren't suppose to change it.

File: docs/faq.rst (hunk 3 above)
Comment (stsewd) [reply docs/faq.rst]: I was thinking about that, even automatically doing it. Maybe a warning/message at the top of the form is better.

File: readthedocs/builds/version_slug.py (hunk 8 above)
Comment (stsewd) [reply readthedocs/builds/version_slug.py]: So, I added it bc it was weird to don't do anything when the slug wasn't right, let me see if I can change the other code to return None for an invalid slug

File: readthedocs/templates/projects/project_version_detail.html [hunk 10]
```diff
@@ -15,7 +15,7 @@
 {% block content-header %}<h1>{% blocktrans with version.name as version_name %}{{ version_name }}{% endblocktrans %}</h1>{% endblock %}
 
 {% block content %}
-    <h2> Editing {{ version.slug }} </h2>
+    <h2> Editing {{ version.verbose_name }} </h2>
```
Comment (stsewd) [inline readthedocs/templates/projects/project_version_detail.html]: I changed this so the user can see the _original_ name of the version. They already see the slug now.

File: readthedocs/templates/projects/project_version_detail.html [hunk 11]
```diff
@@ -31,6 +31,14 @@ <h2> Editing {{ version.slug }} </h2>
 
 
     <form method="post" action=".">
+      <p class="empty">
+        {% url 'projects_redirects' project.slug as redirects_url %}
+        {% blocktrans with redirects_url=redirects_url current_slug=version.slug %}
+          <strong>Warning!</strong> Changing the slug will break existing URLs.
+          If you change it, you may want to <a href="{{ redirects_url }}">create an exact redirect</a>
+          from <span>/en/{{ current_slug }}/$rest</span> to <span>/en/new-slug</span>.
```
Comment (stsewd) [inline readthedocs/templates/projects/project_version_detail.html]: Not really sure if we want to make this _more real_, using current lang, single version, subproject, etc

**Comment by stsewd** [comment]
> I suppose this PR addresses only this part from #5318

Yeah, before automating anything, it should be possible to make it manually

This is how it looks with the new changes

Screenshot_2019-10-02 1 x - rtd-build-tests Read the Docs

File: readthedocs/builds/forms.py [hunk 12]
```diff
@@ -33,10 +33,17 @@ def clean_slug(self):
...
-        if not slugifier.is_valid(slug):
-            msg = _('The slug "{slug}" is not valid.')
-            raise forms.ValidationError(msg.format(slug=original_slug))
+        slug = slugifier.slugify(original_slug)
+        if not slug:
+            ok_chars = ', '.join(VERSION_OK_CHARS)
+            msg = _(
+                'The slug "{slug}" is not valid. '
+                'It should only contain letters, numbers or {ok_chars}. '
+                'And can not start with {ok_chars}.'
```
Comment (humitos) [inline readthedocs/builds/forms.py]:
```suggestion
                'It can not start with {ok_chars}.'
```

File: readthedocs/builds/forms.py [hunk 13]
```diff
@@ -33,10 +33,17 @@ def clean_slug(self):
...
-        slug = slugifier.slugify(original_slug, check_pattern=False)
-        if not slugifier.is_valid(slug):
-            msg = _('The slug "{slug}" is not valid.')
-            raise forms.ValidationError(msg.format(slug=original_slug))
+        slug = slugifier.slugify(original_slug)
+        if not slug:
+            ok_chars = ', '.join(VERSION_OK_CHARS)
+            msg = _(
+                'The slug "{slug}" is not valid. '
+                'It should only contain letters, numbers or {ok_chars}. '
```
Comment (humitos) [inline readthedocs/builds/forms.py]:
```suggestion
                'It should only contain letters, numbers and {ok_chars}. '
```

File: readthedocs/templates/projects/project_version_detail.html (hunk 11 above)
Comment (humitos) [reply readthedocs/templates/projects/project_version_detail.html]: I think it's good enough.

**Comment by stsewd** [comment]
@humitos if this is merged, API v3 should be updated to support this too.

**Comment by ericholscher** [comment]
I'm a little worried that this is a confusing change. It allows users to change slugs, but not verbose_name? How exactly are we using verbose_name and slug now? Is slug now actually an alias, not a slug?

File: readthedocs/builds/version_slug.py [hunk 14]
```diff
@@ -114,13 +104,40 @@ def slugify(self, content):
...
+
+        if not self.is_valid(slug):
+            return None
+        return slug
+
 
-        if not slugified:
-            return self.fallback_slug
-        return slugified
+class VersionSlugField(models.CharField):
```
Comment (ericholscher) [inline readthedocs/builds/version_slug.py]: Why did these files change? There's nothing in the PR description about what logic changes are here.

**Comment by humitos** [comment]
@ericholscher raised a good point today: what has changed from https://github.com/readthedocs/readthedocs.org/pull/4505 that we now decided that we want to support changing a version slug?

I think the main reason is that our code needs to support this because of https://github.com/readthedocs/readthedocs.org/issues/5318 --although, we may don't want to expose this feature to users in a simple form.

I'm not sure, but it may be helpful to review the discussion in the original PR from @davidfischer and write a design document that goes through the benefits and the problems (together with their solutions).

**Comment by stsewd** [comment]
There is a design document that needs review first https://github.com/readthedocs/readthedocs.org/pull/6273
